File: backend/internal/repository/recruitment.go
package repository

import (
	"context"
	"time"

	"hr-saas/internal/db"

	"github.com/google/uuid"
	"github.com/jackc/pgx/v5"
	"github.com/jackc/pgx/v5/pgxpool"
)

type JobPosting struct {
	ID             uuid.UUID `json:"id"`
	TenantID       uuid.UUID `json:"tenant_id"`
	Title          string    `json:"title"`
	Department     string    `json:"department"`
	Location       string    `json:"location"`
	EmploymentType string    `json:"employment_type"`
	Description    string    `json:"description"`
	Requirements   string    `json:"requirements"`
	Status         string    `json:"status"`
	CreatedAt      time.Time `json:"created_at"`
	UpdatedAt      time.Time `json:"updated_at"`
}

type Candidate struct {
	ID          uuid.UUID `json:"id"`
	TenantID    uuid.UUID `json:"tenant_id"`
	JobID       uuid.UUID `json:"job_id"`
	Name        string    `json:"name"`
	Email       string    `json:"email"`
	Phone       string    `json:"phone"`
	ResumeURL   string    `json:"resume_url"`
	Status      string    `json:"status"`
	Source      string    `json:"source"`
	AppliedDate time.Time `json:"applied_date"`
	CreatedAt   time.Time `json:"created_at"`
	UpdatedAt   time.Time `json:"updated_at"`
	// Joins
	JobTitle string `json:"job_title,omitempty"`
}

type Interview struct {
	ID              uuid.UUID `json:"id"`
	TenantID        uuid.UUID `json:"tenant_id"`
	CandidateID     uuid.UUID `json:"candidate_id"`
	InterviewerID   uuid.UUID `json:"interviewer_id"`
	ScheduledAt     time.Time `json:"scheduled_at"`
	DurationMinutes int       `json:"duration_minutes"`
	Location        string    `json:"location"`
	MeetingLink     string    `json:"meeting_link"`
	Status          string    `json:"status"`
	Feedback        *string   `json:"feedback"`
	Rating          *int      `json:"rating"`
	CreatedAt       time.Time `json:"created_at"`
	UpdatedAt       time.Time `json:"updated_at"`
	// Joins
	CandidateName   string `json:"candidate_name,omitempty"`
	InterviewerName string `json:"interviewer_name,omitempty"`
}

// ---------------------------------------------------------
// JOB POSTINGS
// ---------------------------------------------------------

func ListJobPostings(ctx context.Context, pool *pgxpool.Pool, tenantID string) ([]JobPosting, error) {
	var list []JobPosting
	err := db.WithTenant(ctx, pool, tenantID, func(ctx context.Context, tx pgx.Tx) error {
		rows, err := tx.Query(ctx,
			`SELECT id, tenant_id, title, department, location, employment_type, description, requirements, status, created_at, updated_at
			 FROM job_postings
			 ORDER BY created_at DESC`)
		if err != nil {
			return err
		}
		defer rows.Close()
		for rows.Next() {
			var r JobPosting
			if err := rows.Scan(&r.ID, &r.TenantID, &r.Title, &r.Department, &r.Location, &r.EmploymentType, &r.Description, &r.Requirements, &r.Status, &r.CreatedAt, &r.UpdatedAt); err != nil {
				return err
			}
			list = append(list, r)
		}
		return rows.Err()
	})
	return list, err
}

func CreateJobPosting(ctx context.Context, pool *pgxpool.Pool, tenantID uuid.UUID, jp *JobPosting) (*JobPosting, error) {
	var created JobPosting
	err := db.WithTenant(ctx, pool, tenantID.String(), func(ctx context.Context, tx pgx.Tx) error {
		return tx.QueryRow(ctx,
			`INSERT INTO job_postings (tenant_id, title, department, location, employment_type, description, requirements, status)
			 VALUES ($1, $2, $3, $4, $5, $6, $7, COALESCE($8, 'open'))
			 RETURNING id, tenant_id, title, department, location, employment_type, description, requirements, status, created_at, updated_at`,
			tenantID, jp.Title, jp.Department, jp.Location, jp.EmploymentType, jp.Description, jp.Requirements, jp.Status,
		).Scan(&created.ID, &created.TenantID, &created.Title, &created.Department, &created.Location, &created.EmploymentType, &created.Description, &created.Requirements, &created.Status, &created.CreatedAt, &created.UpdatedAt)
	})
	return &created, err
}

// ---------------------------------------------------------
// CANDIDATES
// ---------------------------------------------------------

func ListCandidates(ctx context.Context, pool *pgxpool.Pool, tenantID string) ([]Candidate, error) {
	var list []Candidate
	err := db.WithTenant(ctx, pool, tenantID, func(ctx context.Context, tx pgx.Tx) error {
		rows, err := tx.Query(ctx,
			`SELECT c.id, c.tenant_id, c.job_id, c.name, c.email, c.phone, c.resume_url, c.status, c.source, c.applied_date, c.created_at, c.updated_at,
			        jp.title as job_title
			 FROM candidates c
			 LEFT JOIN job_postings jp ON c.job_id = jp.id
			 ORDER BY c.created_at DESC`)
		if err != nil {
			return err
		}
		defer rows.Close()
		for rows.Next() {
			var r Candidate
			if err := rows.Scan(&r.ID, &r.TenantID, &r.JobID, &r.Name, &r.Email, &r.Phone, &r.ResumeURL, &r.Status, &r.Source, &r.AppliedDate, &r.CreatedAt, &r.UpdatedAt, &r.JobTitle); err != nil {
				return err
			}
			list = append(list, r)
		}
		return rows.Err()
	})
	return list, err
}

// ---------------------------------------------------------
// INTERVIEWS
// ---------------------------------------------------------

func ListInterviews(ctx context.Context, pool *pgxpool.Pool, tenantID string) ([]Interview, error) {
	var list []Interview
	err := db.WithTenant(ctx, pool, tenantID, func(ctx context.Context, tx pgx.Tx) error {
		rows, err := tx.Query(ctx,
			`SELECT i.id, i.tenant_id, i.candidate_id, i.interviewer_id, i.scheduled_at, i.duration_minutes, i.location, i.meeting_link, i.status, i.feedback, i.rating, i.created_at, i.updated_at,
			        c.name as candidate_name, e.name as interviewer_name
			 FROM interviews i
			 JOIN candidates c ON i.candidate_id = c.id
			 LEFT JOIN employees e ON i.interviewer_id = e.id
			 ORDER BY i.scheduled_at DESC`)
		if err != nil {
			return err
		}
		defer rows.Close()
		for rows.Next() {
			var r Interview
			if err := rows.Scan(&r.ID, &r.TenantID, &r.CandidateID, &r.InterviewerID, &r.ScheduledAt, &r.DurationMinutes, &r.Location, &r.MeetingLink, &r.Status, &r.Feedback, &r.Rating, &r.CreatedAt, &r.UpdatedAt, &r.CandidateName, &r.InterviewerName); err != nil {
				return err
			}
			list = append(list, r)
		}
		return rows.Err()
	})
	return list, err
}

File: frontend/lib/data/models/recruitment.dart
class JobPosting {
  final String id;
  final String tenantId;
  final String title;
  final String department;
  final String location;
  final String employmentType;
  final String description;
  final String requirements;
  final String status;
  final DateTime createdAt;
  final DateTime updatedAt;

  JobPosting({
    required this.id,
    required this.tenantId,
    required this.title,
    required this.department,
    required this.location,
    required this.employmentType,
    required this.description,
    required this.requirements,
    required this.status,
    required this.createdAt,
    required this.updatedAt,
  });

  factory JobPosting.fromJson(Map<String, dynamic> json) {
    return JobPosting(
      id: json['id'] as String,
      tenantId: json['tenant_id'] as String,
      title: json['title'] as String,
      department: json['department'] as String? ?? '',
      location: json['location'] as String? ?? '',
      employmentType: json['employment_type'] as String? ?? 'Full-time',
      description: json['description'] as String? ?? '',
      requirements: json['requirements'] as String? ?? '',
      status: json['status'] as String? ?? 'open',
      createdAt: DateTime.parse(json['created_at'] as String),
      updatedAt: DateTime.parse(json['updated_at'] as String),
    );
  }

  Map<String, dynamic> toJson() => {
        'id': id,
        'tenant_id': tenantId,
        'title': title,
        'department': department,
        'location': location,
        'employment_type': employmentType,
        'description': description,
        'requirements': requirements,
        'status': status,
        'created_at': createdAt.toIso8601String(),
        'updated_at': updatedAt.toIso8601String(),
      };
}

class Candidate {
  final String id;
  final String tenantId;
  final String jobId;
  final String name;
  final String email;
  final String phone;
  final String resumeUrl;
  final String status;
  final String source;
  final DateTime appliedDate;
  final DateTime createdAt;
  final DateTime updatedAt;
  // Joined field
  final String? jobTitle;

  Candidate({
    required this.id,
    required this.tenantId,
    required this.jobId,
    required this.name,
    required this.email,
    required this.phone,
    required this.resumeUrl,
    required this.status,
    required this.source,
    required this.appliedDate,
    required this.createdAt,
    required this.updatedAt,
    this.jobTitle,
  });

  factory Candidate.fromJson(Map<String, dynamic> json) {
    return Candidate(
      id: json['id'] as String,
      tenantId: json['tenant_id'] as String,
      jobId: json['job_id'] as String? ?? '',
      name: json['name'] as String,
      email: json['email'] as String,
      phone: json['phone'] as String? ?? '',
      resumeUrl: json['resume_url'] as String? ?? '',
      status: json['status'] as String? ?? 'new',
      source: json['source'] as String? ?? '',
      appliedDate: json['applied_date'] != null ? DateTime.parse(json['applied_date'] as String) : DateTime.now(),
      createdAt: DateTime.parse(json['created_at'] as String),
      updatedAt: DateTime.parse(json['updated_at'] as String),
      jobTitle: json['job_title'] as String?,
    );
  }
}

class Interview {
  final String id;
  final String tenantId;
  final String candidateId;
  final String interviewerId;
  final DateTime scheduledAt;
  final int durationMinutes;
  final String location;
  final String meetingLink;
  final String status;
  final String? feedback;
  final int? rating;
  final DateTime createdAt;
  final DateTime updatedAt;
  // Joined fields
  final String? candidateName;
  final String? interviewerName;

  Interview({
    required this.id,
    required this.tenantId,
    required this.candidateId,
    required this.interviewerId,
    required this.scheduledAt,
    required this.durationMinutes,
    required this.location,
    required this.meetingLink,
    required this.status,
    this.feedback,
    this.rating,
    required this.createdAt,
    required this.updatedAt,
    this.candidateName,
    this.interviewerName,
  });

  factory Interview.fromJson(Map<String, dynamic> json) {
    return Interview(
      id: json['id'] as String,
      tenantId: json['tenant_id'] as String,
      candidateId: json['candidate_id'] as String,
      interviewerId: json['interviewer_id'] as String? ?? '',
      scheduledAt: DateTime.parse(json['scheduled_at'] as String),
      durationMinutes: json['duration_minutes'] as int? ?? 30,
      location: json['location'] as String? ?? '',
      meetingLink: json['meeting_link'] as String? ?? '',
      status: json['status'] as String? ?? 'scheduled',
      feedback: json['feedback'] as String?,
      rating: json['rating'] as int?,
      createdAt: DateTime.parse(json['created_at'] as String),
      updatedAt: DateTime.parse(json['updated_at'] as String),
      candidateName: json['candidate_name'] as String?,
      interviewerName: json['interviewer_name'] as String?,
    );
  }
}
